wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java

```diff
@@ -197,26 +194,49 @@ else if (isNewPageInstance() == false)
 	}
 
 	/**
-	 * The page instance is new only if there is no cached instance or the data stores doesn't have
-	 * a page with that id with the same {@linkplain #pageClass}.
-	 * 
-	 * @see IPageProvider#isNewPageInstance()
+	 * @return negates {@link PageProvider#hasPageInstance()}
+	 * @deprecated use {@link PageProvider#hasPageInstance()} negation instead
 	 */
 	@Override
 	public boolean isNewPageInstance()
 	{
-		boolean isNew = pageInstance == null;
-		if (isNew && pageId != null)
+		return !hasPageInstance();
+	}
+
+	/**
+	 * If this provider returns existing page, regardless if it was already created by PageProvider
+	 * itself or is or can be found in the data store. The only guarantee is that by calling
+	 * {@link PageProvider#getPageInstance()} this provider will return an existing instance and no
+	 * page will be created.
+	 * 
+	 * @return if provides an existing page
+	 */
+	@Override
+	public final boolean hasPageInstance()
+	{
+		if (provision != null || pageId != null)
 		{
-			IRequestablePage storedPageInstance = getStoredPage(pageId);
-			if (storedPageInstance != null)
-			{
-				pageInstance = storedPageInstance;
-				isNew = false;
-			}
+			return getProvision().didResolveToPage();
 		}
+		else
+			return false;
+	}
 
-		return isNew;
+	/**
+	 * Returns whether or not the page instance held by this provider has been instantiated by the
+	 * provider.
+	 * 
+	 * @return {@code true} iff the page instance held by this provider was instantiated by the
+	 *         provider
+	 */
+	@Override
+	public final boolean doesProvideNewPage()
+	{
+		if (provision == null)
+		{
+			throw new IllegalStateException("Page instance not yet resolved");
```

Comment:
This doesn't look nice.
At https://issues.apache.org/jira/browse/WICKET-4201?focusedCommentId=15853108&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15853108 you say 
 ` removed the PageExpiredException from PageProvider since it breaks the class coherence. It's designed to provide pages and related metadata (such as if the page was expired included), not to change the application flow on application exceptions.` but this runtime exception does exactly this.
IMO it would be better to return a three state result: `NEW`, `OLD` and `UNKNOWN`